### PR TITLE
linker: add support for odm partition

### PR DESCRIPTION
**Pull request by sledges**
Android O Treble project moves proprietary vendor binaries to a separate /odm
partition.

Manufacturers like Sony start moving to the new layout as early as possible [1][2],
hence we need to adapt libhybris as well.

[1] https://android.googlesource.com/platform/bionic/+/0e093b770af71a9da6e6556ce6f2bda8ba4fb40b%5E%21/#F0
[2] https://android.googlesource.com/platform/bionic/+/645587e5590f5ca42dc3ceecd0085cc446e92ccb%5E!/#F0

### CODE REVIEW

File: hybris/common/n/linker.cpp [hunk 1]
```diff
@@ -2507,6 +2514,12 @@ void* do_dlopen(const char* name, int flags, const android_dlextinfo* extinfo,
         translated_name = asan_name_holder.c_str();
         PRINT("linker_asan dlopen translating \"%s\" -> \"%s\"", name, translated_name);
       }
+    } else if (file_is_in_dir(name, kOdmLibDir)) {
+      asan_name_holder = std::string(kAsanOdmLibDir) + "/" + basename(name);
+      if (file_exists(asan_name_holder.c_str())) {
+        translated_name = asan_name_holder.c_str();
+        PRINT("linker_asan dlopen translating \"%s\" -> \"%s\"", name, translated_name);
+      }
```
Comment (sledges) [inline hybris/common/n/linker.cpp:2522]: This would need refactoring, but it's done in the O linker

File: hybris/common/mm/linker.cpp [hunk 2]
```diff
@@ -83,9 +83,11 @@ static soinfo* somain; // main process, always the one after libdl_info
 
 static const char* const kDefaultLdPaths[] = {
 #if defined(__LP64__)
+  "/odm/lib64",
```
Comment (krnlyng) [inline hybris/common/mm/linker.cpp]: Here /odm is first

File: hybris/common/n/linker.cpp [hunk 3]
```diff
@@ -169,18 +169,23 @@ static soinfo* somain; // main process, always the one after libdl_info
...
+static const char* const kOdmLibDir        = "/odm/lib";
 static const char* const kAsanSystemLibDir = "/data/lib";
 static const char* const kAsanVendorLibDir = "/data/vendor/lib";
+static const char* const kAsanOdmLibDir    = "/data/odm/lib";
 #endif
 
 static const char* const kDefaultLdPaths[] = {
   kSystemLibDir,
   kVendorLibDir,
+  kOdmLibDir,
```
Comment (krnlyng) [inline hybris/common/n/linker.cpp:188]: Here it is last

File: hybris/configure.ac [hunk 4]
```diff
@@ -137,9 +137,9 @@ AC_ARG_WITH(default-egl-platform,
 AC_SUBST(DEFAULT_EGL_PLATFORM)
 
 if test "x$arch" = "xarm" -o "x$arch" = "xx86"; then
-  DEFAULT_HYBRIS_LD_LIBRARY_PATH="/vendor/lib:/system/lib"
+  DEFAULT_HYBRIS_LD_LIBRARY_PATH="/odm/lib:/vendor/lib:/system/lib"
```
Comment (krnlyng) [inline hybris/configure.ac]: Here it is first...

Why?

File: hybris/common/mm/linker.cpp (hunk 2 above)
Comment (sledges) [reply hybris/common/mm/linker.cpp]: https://android.googlesource.com/platform/bionic/+/f74b041cce75091b922151d4fa57e63c8a6caf4b%5E!/

File: hybris/common/n/linker.cpp (hunk 3 above)
Comment (sledges) [reply hybris/common/n/linker.cpp:188]: https://android.googlesource.com/platform/bionic/+/0e093b770af71a9da6e6556ce6f2bda8ba4fb40b%5E%21/#F0

File: hybris/configure.ac (hunk 4 above)
Comment (sledges) [reply hybris/configure.ac]: Go figure :D 

I picked the first precedent (from M)

**Comment by krnlyng** [comment]
LGTM

**Pull request merged**